interfaces/prompting/requestprompts/requestprompts.go

```diff
@@ -0,0 +1,384 @@
+func (pdb *PromptDB) PromptWithID(user uint32, id string) (*Prompt, error) {
+	pdb.mutex.Lock()
+	defer pdb.mutex.Unlock()
+	_, prompt, err := pdb.promptWithID(user, id)
+	return prompt, err
+}
+
+// promptWithID returns the user entry for the given user and the prompt with
+// the given ID for the that user.
+//
+// The caller should hold the prompt DB lock.
+func (pdb *PromptDB) promptWithID(user uint32, id string) (*userPromptDB, *Prompt, error) {
+	userEntry, ok := pdb.perUser[user]
+	if !ok || len(userEntry.ByID) == 0 {
+		return nil, nil, ErrNotFound
+	}
+	prompt, ok := userEntry.ByID[id]
+	if !ok {
+		return nil, nil, ErrNotFound
+	}
+	return userEntry, prompt, nil
+}
+
+// Reply resolves the prompt with the given ID using the given outcome by
+// sending a reply to all associated listener requests, then removing the
+// prompt from the prompt DB.
+//
+// Records a notice for the prompt, and returns the prompt's former contents.
+func (pdb *PromptDB) Reply(user uint32, id string, outcome prompting.OutcomeType) (*Prompt, error) {
+	allow, err := outcome.AsBool()
+	if err != nil {
+		return nil, err
+	}
+	pdb.mutex.Lock()
+	defer pdb.mutex.Unlock()
+	userEntry, prompt, err := pdb.promptWithID(user, id)
+	if err != nil {
+		return nil, err
+	}
+	for _, listenerReq := range prompt.listenerReqs {
+		if err := sendReply(listenerReq, allow); err != nil {
+			return nil, err
+		}
+	}
+	delete(userEntry.ByID, id)
+	data := make(map[string]string)
+	data["resolved"] = "replied"
+	pdb.notifyPrompt(user, id, data)
+	return prompt, nil
+}
+
+var sendReply = func(listenerReq *listener.Request, reply interface{}) error {
+	return listenerReq.Reply(reply)
+}
+
+// HandleNewRule checks if any existing prompts are satisfied by the given rule
+// contents and, if so, sends back a decision to their listener requests.
+//
+// A prompt is satisfied by the given rule contents if the user, snap,
+// interface, and path of the prompt match those of the rule, and if either the
+// outcome is "allow" and all of the prompt's permissions are matched by those
+// of the rule contents, or if the outcome is "deny" and any of the permissions
+// match.
+//
+// Records a notice for any prompt which was satisfied, or which had some of
+// its permissions satisfied by the rule contents. In the future, only the
+// remaining unsatisfied permissions of a partially-satisfied prompt must be
+// satisfied for the prompt as a whole to be satisfied.
+//
+// Returns the IDs of any prompts which were fully satisfied by the given rule
+// contents.
+func (pdb *PromptDB) HandleNewRule(metadata *prompting.Metadata, constraints *prompting.Constraints, outcome prompting.OutcomeType) ([]string, error) {
+	allow, err := outcome.AsBool()
+	if err != nil {
+		return nil, err
+	}
+	pdb.mutex.Lock()
+	defer pdb.mutex.Unlock()
+	satisfiedPromptIDs := []string{}
+	userEntry, ok := pdb.perUser[metadata.User]
+	if !ok {
+		return satisfiedPromptIDs, nil
```

Comment:
Can you just return nil here?

Reply:
No, for the same reason as above. We want to marshal as `"[]"` instead of `null`.

Reply:
True, better for the daemon to handle this rather than assuming it of the functions it's calling. I'll switch to `nil` when it's easier/clearer than `[]string{}`.